### PR TITLE
Enable Azure Workload Identity for `fsspec` in `flytekit`

### PR DESCRIPTION
**Pull request by fiedlerNr9**
# TL;DR
This change enables flytekit to make use of aks enabled workload identities. Before this change, reading and writing data to a storage account was only possible with setting env variables for the storage account key. With setting `anon` to False, the Default Credentials, which are provides by aks workload identities can be used. [See](https://github.com/fsspec/adlfs/blob/main/adlfs/spec.py#L193)

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 I tried the change myself by following this [guide](https://learn.microsoft.com/en-us/azure/aks/learn/tutorial-kubernetes-workload-identity). IMO, using workload identities should be default on azure since when connecting to other Azure Services like Key Vaults, workload identities are the way to go as well. 

## Tracking Issue
[3962](https://github.com/flyteorg/flyte/issues/3962)

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/data_persistence.py [hunk 1]
```diff
@@ -119,6 +119,9 @@ def get_filesystem(
             if anonymous:
                 kwargs["token"] = _ANON
             return fsspec.filesystem(protocol, **kwargs)  # type: ignore
+        elif protocol == "abfs":
+            kwargs["anon"] = False
```
[the remark this answers is not in the document]
Comment (fiedlerNr9) [reply flytekit/core/data_persistence.py:123]: It really looks like that the purpose of `anon = True` is to disable using Azure DefaultCredentials in [adlfs](https://github.com/fsspec/adlfs/blob/main/adlfs/spec.py#L282-L286). Even with always `anon = False` you can still connect via storage account SAS token or storage account key. 

Would you suggest to set:  `kwargs["anon"] = anonymous`?

**Review by wild-endeavor — COMMENTED** [review]
@fiedlerNr9 were you able to test this against a real azure bucket?  like run a workflow with this pr installed in the image.  let's merge if that worked.  thank you.  we can find a more general solution in the future.

**Comment by fiedlerNr9** [comment]
@wild-endeavor Yes i did and it works like expected.

image
image

One can see that the flyte pods now authenticate with `AZURE_FEDERATED_TOKEN_FILE` to the azure storage account

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
